### PR TITLE
Changes

### PR DESCRIPTION
**Pull request by StrawWagen**
Removed DMR, Rifle, and Anti-Mat turret
Turrets now target shaped charges

Rebalanced all turrets
ent_jack_turret_mg
made it op for holding medium long hallways and that only, 
dps rivals shotgun turret
easy to get ahead of it's aiming
takes lots and lots of power

ent_jack_turret_smg
good at covering wide fov, medium short range, 
good dps
hard to get ahead of it's aiming
takes lots of power

plinker and pistol are just worse versions of the smg, but they don't take much power, can place and forget.

ent_jack_turret_shotty
short range
instant turning
one shot kill 100% of the time
super high powerdraw

ent_jack_turret_sniper
made into autosniper 
medium-slow shooting, slow turning, long range
4 shots to kill
high powerdraw

ent_jack_teslasentry
generally more useful, less fragile.
zaps THROUGH metal props
zaps all default projectiles out of the sky
sucks up combine balls, can lead to explosive overcharge

+ lots of minor QOL changes not mentioned

### CODE REVIEW

File: lua/entities/ent_jack_turret_base/init.lua [hunk 1]
```diff
@@ -102,7 +112,15 @@ function ENT:ExternalCharge( amt )
...
 
 function ENT:Initialize()
-    self:SetAngles( Angle( 0, 0, 0 ) )
+
+    local spawnAng = Angle( 0, 0, 0 )
+    local owner = self:GetNWEntity( "Owner", nil )
+    if owner then
+        spawnAng.y = owner:EyeAngles().y
+
+    end
```
Comment (StrawWagen) [inline lua/entities/ent_jack_turret_base/init.lua]:
```suggestion
    local owner = self:GetNWEntity( "Owner", nil )
    if IsValid( owner ) then
        spawnAng.y = owner:EyeAngles().y

    end

```

**Review by StrawWagen — COMMENTED** [review]
add IsValid

File: lua/entities/ent_jack_40mmgrenade/init.lua [hunk 2]
```diff
@@ -44,20 +44,22 @@ function ENT:Detonate()
...
 
-    util.BlastDamage( self, attacker, pos, 190, 190 )
+    util.BlastDamage( self, attacker, pos, owner.BulletDamage, owner.BulletDamage )
 
     local plooie = EffectData()
     plooie:SetOrigin( self:GetPos() )
     plooie:SetScale( .75 )
     plooie:SetRadius( 2 )
-    plooie:SetNormal( self:GetForward() )
+    plooie:SetNormal( Vector( 0,0,1 ) )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_40mmgrenade/init.lua]: Creating vectors on the fly is slow. Could you assign this to a top level variable? _(also, spaces after commas)_

File: lua/entities/ent_jack_teslasentry/cl_init.lua [hunk 3]
```diff
@@ -89,20 +89,17 @@ end
...
+        if rag:GetClass() ~= "class C_ClientRagdoll" then continue end
+        for i = 1, 60 do
+            timer.Simple( i / 20, function()
+                if not IsValid( rag ) then return end
+                local Bones = rag:GetPhysicsObjectCount() - 1
+                local Obj = rag:GetPhysicsObjectNum( math.random( 2, Bones ) )
+
+                if not Obj then return end
+                Obj:ApplyForceCenter( VectorRand() * ( 60 - i ) * Obj:GetMass() * 10 )
+            end )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_teslasentry/cl_init.lua]: Can we just remove this whole method for now? This is truly awful stuff (not your fault)
We'd also need to remove all of the usermessage stuff for `JackysElectriTwitchClientSentry` (both the sending and receiving)

File: lua/entities/ent_jack_turret_amateriel/shared.lua [hunk 4]
```diff
@@ -2,6 +2,6 @@ ENT.Type = "anim"
 ENT.PrintName = "Anti-Materiel ST"
 ENT.Author = "Jackarunda"
 ENT.Category = "J.I. Defense Solutions"
-ENT.Spawnable = true
+ENT.Spawnable = false
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_amateriel/shared.lua]: If we're getting rid of this for good, just delete it all. The materials, sounds, lua fules - all of it

File: lua/entities/ent_jack_turret_base/init.lua [hunk 5]
```diff
@@ -51,6 +51,7 @@ ENT.RoundsOnBelt = 0
 ENT.NextWarnTime = 0
 ENT.WillWarn = false
 ENT.WillLight = false
+ENT.WillLightOverride = nil
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua:54]: Is this overriding a not-nil value? If not, you can remove this line

File: lua/entities/ent_jack_turret_base/init.lua [hunk 6]
```diff
@@ -102,7 +112,16 @@ function ENT:ExternalCharge( amt )
 end
 
 function ENT:Initialize()
-    self:SetAngles( Angle( 0, 0, 0 ) )
+
+    local spawnAng = Angle( 0, 0, 0 )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua]: Can you move this var outside of this function?

File: lua/entities/ent_jack_turret_base/init.lua [hunk 7]
```diff
@@ -616,10 +646,13 @@ function ENT:Traverse()
     end
 
     if IsValid( self.flashlight ) then
-        self.flashlight:SetLocalAngles( self:WorldToLocalAngles( self:GetAttachment( 1 ).Ang ) )
+        self.flashlight:SetLocalAngles( Angle( self.CurrentSwing, self.CurrentSweep, 0 ) )
     end
 end
 
+function ENT:AdditionalShootFX()
+end
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua:655]: Can you add a comment explaining why this was emptied?

File: lua/entities/ent_jack_turret_base/init.lua [hunk 8]
```diff
@@ -690,15 +725,28 @@ function ENT:FireShot()
...
-        Src = SelfPos
+        Spread = SpreadVec,
+        Src = SelfPos,
+        Callback = function( _, trace )
+            local TracerEffect = EffectData()
+            TracerEffect:SetStart( SelfPos )
+            TracerEffect:SetOrigin( trace.HitPos )
+            TracerEffect:SetFlags( 1 )
+            TracerEffect:SetScale( 30000 ) -- usain bolt speed
+            util.Effect( self.TracerEffect, TracerEffect ) -- BIG effect
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua]:
```suggestion
            TracerEffect:SetScale( 30000 )
            util.Effect( self.TracerEffect, TracerEffect )

```

File: lua/entities/ent_jack_turret_base/init.lua [hunk 9]
```diff
@@ -690,15 +725,28 @@ function ENT:FireShot()
         end
     end
 
+    local SpreadVec = Vector() -- manually build spread to avoid weird bug where all shots land in the top right for no apparent reason
+
+    SpreadVec.x = spread * math.Rand( -1, 1 )
+    SpreadVec.y = spread * math.Rand( -1, 1 )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua]:
```suggestion
    local SpreadVec = Vector( spread, spread, spread ) + VectorRand( -1, 1 )
```
you can multiply the `VectorRand` by some `intensity` amount I think

File: lua/entities/ent_jack_turret_base/init.lua [hunk 10]
```diff
@@ -788,17 +838,31 @@ function ENT:StandBy()
...
+        return true
+    elseif allowableThinnessSqr and not traceResult.HitWorld and traceResult.Hit then
+        if traceResult.HitPos:DistToSqr( worldSpaceCenter ) < allowableThinnessSqr then
+            return true
+        else
+            return false
+        end
+    else
+        return false
+    end
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua]:
```suggestion
    if not traceResult.Hit then return true end
    
    if not allowableThinnessSqr then return false end
    if traceResult.HitWorld then return false end

    return traceResult.HitPos:DistToSqr( worldSpaceCenter ) < allowableThinnessSqr
```

File: lua/entities/ent_jack_turret_dmr/shared.lua [hunk 11]
```diff
@@ -2,6 +2,6 @@ ENT.Type = "anim"
 ENT.PrintName = "Marksman ST"
 ENT.Author = "Jackarunda"
 ENT.Category = "J.I. Defense Solutions"
-ENT.Spawnable = true
+ENT.Spawnable = false
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_dmr/shared.lua]: Same comment about removing them entirely

File: lua/entities/ent_jack_turret_pistol/init.lua [hunk 12]
```diff
@@ -19,12 +20,13 @@ ENT.NearShotNoise = "snd_jack_turretshootshort_close.mp3"
 ENT.FarShotNoise = "snd_jack_turretshootshort_far.mp3"
 ENT.AmmoType = "9x19mm"
 ENT.MuzzEff = "muzzleflash_pistol"
-ENT.BarrelSizeMod = Vector( .9, .9, .9 )
+ENT.Automatic = nil
+ENT.BarrelSizeMod = Vector( 0.9, 0.9, 0.9 )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_pistol/init.lua:24]: 👍

File: lua/entities/ent_jack_turret_rifle/shared.lua [hunk 13]
```diff
@@ -2,6 +2,6 @@ ENT.Type = "anim"
 ENT.PrintName = "Rifle ST"
 ENT.Author = "Jackarunda"
 ENT.Category = "J.I. Defense Solutions"
-ENT.Spawnable = true
+ENT.Spawnable = false
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_rifle/shared.lua]: Same comment about removing them entirely

File: lua/entities/ent_jack_turretammobox_338/shared.lua [hunk 14]
```diff
@@ -1,8 +1,7 @@
 ENT.Type 			= "anim"
 ENT.PrintName		= ".338LM AmmoBox"
 ENT.Author			= "Jackarunda"
-ENT.Category			= "J.I. Defense Solutions"
+ENT.Category		= "J.I. Defense Solutions"
+ENT.Spawnable       = false
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turretammobox_338/shared.lua]: Same comment about deleting

**Review by sarahsturgeon — CHANGES_REQUESTED** [review]

File: lua/entities/ent_jack_40mmgrenade/init.lua (hunk 2 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_40mmgrenade/init.lua]: Applied

File: lua/entities/ent_jack_teslasentry/cl_init.lua (hunk 3 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_teslasentry/cl_init.lua]: Applied

File: lua/entities/ent_jack_turret_amateriel/shared.lua (hunk 4 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_amateriel/shared.lua]: Applied, removed the ammo too.

File: lua/entities/ent_jack_turret_base/init.lua (hunk 10 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: Applied

File: lua/entities/ent_jack_turret_rifle/shared.lua (hunk 13 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_rifle/shared.lua]: Applied!

File: lua/entities/ent_jack_turretammobox_338/shared.lua (hunk 14 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turretammobox_338/shared.lua]: Applied

File: lua/entities/ent_jack_turret_base/init.lua (hunk 6 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: Applied!

File: lua/entities/ent_jack_turret_dmr/shared.lua (hunk 11 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_dmr/shared.lua]: Done!

File: lua/entities/ent_jack_turret_base/init.lua (hunk 7 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua:655]: Done!

File: lua/entities/ent_jack_turret_base/init.lua (hunk 8 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: Applied :(

File: lua/entities/ent_jack_turret_base/init.lua (hunk 9 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua]: Applied, now uses VectorRand as the "random creator".
Z is behind/in front, just makes the turrets biast towards shooting in the center.

File: lua/entities/ent_jack_turret_base/init.lua (hunk 5 above)
Comment (StrawWagen) [reply lua/entities/ent_jack_turret_base/init.lua:54]: I put it there so it's with the other configurable variables.

File: lua/entities/ent_jack_40mmgrenade/init.lua [hunk 15]
```diff
@@ -39,25 +39,29 @@ function ENT:Think()
     return true
 end
 
+local vecUp = Vector( 0,0,1 )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_40mmgrenade/init.lua:42]:
```suggestion
local vecUp = Vector( 0, 0, 1 )
```

File: lua/entities/ent_jack_landmine/cl_init.lua [hunk 16]
```diff
@@ -23,7 +23,7 @@ function ENT:Draw()
     end
 
     if planted and not self.ColorApplied then
-        local color = render.GetSurfaceColor( self:GetPos() + Vector( 0, 10, 0 ) + self:GetForward() * 10, self:GetPos() + Vector( 0, 10, 0 ) - self:GetForward() * 10 )
+        local color = render.GetSurfaceColor( self:GetPos() + Vector( 0, 10, 0 ) + self:GetForward() * 10, self:GetPos() + Vector( 0, 10, 0 ) - self:GetForward() * 100 )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_landmine/cl_init.lua:26]: Could you move these vectors out into their own var too

File: lua/entities/ent_jack_turret_base/init.lua [hunk 17]
```diff
@@ -711,6 +763,8 @@ function ENT:FireShot()
     sound.Play( self.NearShotNoise, SelfPos, 75, self.ShootSoundPitch )
     sound.Play( self.FarShotNoise, SelfPos + Vector( 0, 0, 1 ), 90, self.ShootSoundPitch - 10 )
 
+    local Scayul = 1
+
     if self:GetClass() ~= "ent_jack_turret_plinker" then
         sound.Play( self.NearShotNoise, SelfPos, 75, self.ShootSoundPitch )
         sound.Play( self.FarShotNoise, SelfPos + Vector( 0, 0, 1 ), 110, self.ShootSoundPitch - 10 )
```
Comment (sarahsturgeon) [inline lua/entities/ent_jack_turret_base/init.lua:770]: More static vectors that can be extracted

**Review by sarahsturgeon — APPROVED** [review]
LGTM

I left a few comments for code that you didn't add, but that would be great opportunities for improvement. If you want to take them, we can do it later too.

Especially the one in the Draw hook- that one's bad

File: lua/entities/ent_jack_40mmgrenade/init.lua (hunk 15 above)
Comment (periapsises) [reply lua/entities/ent_jack_40mmgrenade/init.lua:42]: Could even use that if you wanna be fancy
```lua
local vecUp = vector_up
```

**Pull request merged**